Performance improvements

mphe opened the pull request:
This PR salvages #122 and adds some slight additional refactoring by caching vertices and tesselated points.
Also implements the code cleanup mentioned in #122.

Notable change: Since points are now computed lazily, but some code needs immediate access to them (like tests), there is now a `force_update()` function.

It might be useful to call `force_update` in `SS2D_Action`s that manipulate points, but not sure if necessary.

limbonaut commented:
With these changes, "defer mesh updates" option doesn't seem to work as it used to.

limbonaut commented:
I think the issue is that because `changed` signal is not emitted until action is fully committed, we need to rebuild the point cache:
```diff
diff --git a/addons/rmsmartshape/shapes/shape.gd b/addons/rmsmartshape/shapes/shape.gd
index c504f33..1c02179 100644
--- a/addons/rmsmartshape/shapes/shape.gd
+++ b/addons/rmsmartshape/shapes/shape.gd
@@ -478,6 +478,12 @@ func is_index_in_range(idx: int) -> bool:
 
 func set_point_position(key: int, pos: Vector2) -> void:
 	_points.set_point_position(key, pos)
+	if _points.is_updating():
+		# Note: With deferred updates option, `changed` signal won't be emitted
+		#       until action is fully commited (user has released the mouse button).
+		# 		So we need to update the tesselation cache here.
+		_vertex_cache[_points.get_point_index(key)] = pos
+		_cache_tesselation()
 
 
 func remove_point(key: int) -> void:

```

mphe commented:
Good idea. I'll look into it tomorrow. I guess tesselation cache can also be moved to PointArray.

limbonaut commented:
> Good idea. I'll look into it tomorrow. I guess tesselation cache can also be moved to PointArray.

Looking at https://github.com/SirRamEsq/SmartShape2D/blob/b585fc7a378fbe68955cb1ab127bc020cbc1c86d/addons/rmsmartshape/shapes/point_array.gd#L334C1-L338C17, we could add `points_changed` signal and rebuilt cache (or both caches) in a handler. And at the same time keep the `changed` signal for mesh updates. I'd rename it, though -- `changed` and `points_changed` are too similar. Maybe `changed -> update_finished` or something like that.

mphe commented:
I'm currently looking into it. You mean emitting another signal when `_updating` is true?
I don't like that idea because it will produce inconsistent results, as it will only work with code using `begin_update()` and `end_update()`.
I prefer the approach of having `SS2D_Point_Array` manage the point cache itself and regenerating it lazily when requested.
This is straight forward and works seamlessly in all cases.

limbonaut commented:
> I'm currently looking into it. You mean emitting another signal when `_updating` is true? I don't like that idea because it will produce inconsistent results, as it will only work with code using `begin_update()` and `end_update()`. I prefer the approach of having `SS2D_Point_Array` manage the point cache itself and regenerating it lazily when requested. This is straight forward and works seamlessly in all cases.

Oh, I meant a different set of signals: one that notifies when the point array is `changed`, and another when `update_finished`. In such setup, `update_finished` signifies the end of updates to the point array by the plugin code (a series of contextually connected operations) and is used to rebuild the mesh, and `changed` would be used to invalidate any data that depends on the point array, like tessellation data and the curve. The difference between the two signals is frequency: `changed` would be emitted any time the point data is changed, and as such could be used for cache invalidation. At this moment, we only have the `update_finished` signal (I mean it semantically ofc - it is called `changed` right now). And it's not emitted any time a change happens, but only when a series of such changes are finalized with the `end_update()` or when `begin_update/end_update` is not used (in case of simple actions).

And I agree that moving vertex cache into `PointArray` is a good idea. We only need it because of how the point array data is constructed and how it misaligns with the rest of the code. But I'm not sure if tessellation should belong to that class.

mphe commented:
Ah, I see. That is actually a good idea regardless of the endeavour of this PR, as it provides a better distinction when related code should run.

I noticed `_curve` also needs to be cached, as it is used for tesselation and other functions.
Now there are four options:
1. Cache vertices, curve and tesselation points in `PointArray`
2. Cache vertices, curve and tesselation points in `Shape`
3. Cache only vertices in `PointArray` and the rest in `Shape`
4. Create a new class (`VertexCache`, `PointProvider`, ...) and move all point caching functionality there.

I can see why `PointArray` should not be responsible for caching curve and tesselation.
On the other hand, `PointArray` also handles curve in/out handles, hence it is not so far-fetched to let it cache the curve and tesselation and provide a central interface for all point related data.
Also these three components kinda belong together semantically.

Regarding refactoring discussions we had in Discord about removing the whole point-array-point-key mapping and using a simple list of `Point` objects instead, it would make sense to cache vertices in `PointArray` and also provide a `get_vertices()´ function there.

Option 4) would decouple point caching from both `Shape` and `PointArray` while keeping the vertices, curve and tesselation together.
I find this idea actually quite clean.
The more I think about it, I find this option best, as caching lies semantically somewhere between `Shape` and `PointArray`.

Regarding option 3) + an `update_finished` signal, `PointArray` would handle vertex cache and `Shape` would handle curve/tesselation cache. `Shape` would listen for `changed` signals to invalidate the internal curve/tesselation cache and rebuild it lazily when the data is requested by `get_tesselation_points` or `get_curve()`.
It also introduces another dirty flag in `Shape` (something like `_points_dirty`) that is only related to keeping the point cache up-to-date but not the complete mesh.

I don't particulary like 3) as it distributes computation of tightly related data across two classes.
Personally, I'd go for 1) or 4), but I'm open for discussion.

Regardless of which option is chosen, this change requires adapting various places where the tesselation and curve cache is accessed directly.

limbonaut commented:
Should also consider the use-case when several shapes share a single `PointArray` resource. The whole reason why we started to rely on the `changed` signal during 4.x development is partly because we retired the node that was responsible for synchronization of the point data between several shapes. Now you can simply copy and paste the shapes to do that - and they reuse the resource. I think **1)** would allow reusing this cached data instead of recalculating it in each shape instance that uses the same point array resource. And 1 and 4 can also be combined, with `PointArray` handling the requests and delegating the cache calculation to a specialized class (or classes). I don't think we need a `VertexCache` though - after all, it's just an array of points and is the domain of `PointArray` class (in code, `_vertex_cache` is just a plain array of all points in their proper key order).

limbonaut commented on addons/rmsmartshape/lib/TesselationVertexMapping.gd:
```diff
@@ -0,0 +1,35 @@
+extends RefCounted
```
Better stick with the same file naming convention as in the rest of the code base, which is snake_case.

limbonaut commented:
I gave it a spin, and it all works great - I didn't find any regressions. Aside from a couple of nitpicks above, this PR looks good to me. And thanks for adding the doc comments in various places!

mphe commented:
I wasn't completely finished, yet, I'm still working on the `update_finished` signal but couldn't find the time, yet.

But since we're here already, here are some notes to the changes so far:
- Moved all point related caching to PointArray, including the new tesselation <-> vertex index mapping
- Point cache is now recomputed lazily on access when dirty.
As a result, there is now a separation between point cache and edge cache. Point cache is computed in PointArray, edge cache in `shape.gd`. Point cache is recomputed lazily when dirty, edge cache is recomputed deferred when dirty.
That means, code has seamless access to all point related data, while edge data might need a `shape.force_update()` if the data is required immediately.
- I also deprecated various functions in `shape.gd` that act as a wrapper to PointArray. IMO this is redundant and respective functions should be called directly on the PointArray object.
- I decided against implementing 4) (for now) because I find it clearer without it. There is not that much additional code related to caching that would be worth extracting into a separate class.
An exception is SS2D_TesselationVertexMapping, because it is used directly in a unit test and I wanted to keep it that way (for now).
Maybe the caching functionality should be extracted in a separate class in the future to similarily allow better unit testing, but not sure.
- Most of the doc comments are actually taken from respective wrapper functions in `shape.gd`. I just moved them over to PointArray.

I will try to update in the next few days and fix the mentioned change requests.

mphe commented:
Doesn't make much sense IMO because edges usually need to update every frame to reflect the current state, except when deferred updates are enabled, but that use-case is already handled.
There are also not many cases where outside code needs to access the generated edges.
Thinking about it, we don't really have edge cache as such. What I meant with edge cache was more like edge data generation, i.e. meshes and collisions.

mphe commented:
I have noticed some `.duplicate()` calls in PointArray and in `get_curve()`.

https://github.com/SirRamEsq/SmartShape2D/blob/b585fc7a378fbe68955cb1ab127bc020cbc1c86d/addons/rmsmartshape/shapes/point_array.gd#L160-L171

https://github.com/SirRamEsq/SmartShape2D/blob/b585fc7a378fbe68955cb1ab127bc020cbc1c86d/addons/rmsmartshape/shapes/shape.gd#L206-L207

Theses are just some examples. There are `duplicate()`s everywhere in PointArray.
I wonder if this is actually necessary. Sometimes those functions are never even used in the project.
I can kinda see why a copy would be safer but on the other hand it seems like a needless copy and I would even expect to get the original object in some cases.
PointArray even listens vor changed signals on each point to make sure they can be safely modified directly.
So I think it's safe to remove the duplication and just return the object. What do you think?
As for `get_curve()`, I already removed the `duplicate()` in the `PointArray.get_curve()` but kept it in the now deprecated wrapper in shape.gd for compatibility.

mphe commented on addons/rmsmartshape/shapes/point_array.gd:
```diff
@@ -19,10 +19,30 @@ enum CONSTRAINT { NONE = 0, AXIS_X = 1, AXIS_Y = 2, CONTROL_POINTS = 4, PROPERTI
+## Controls how many degrees the midpoint of a segment may deviate from the real
+## curve, before the segment has to be subdivided.
+@export_range(0.1, 8.0, 0.1, "or_greater", "or_lesser")
+var tessellation_tolerence: float = 4.0 : set = set_tessellation_tolerence
```
Can change the property in PointArray, but not in Shape to maintain compatibility, until we eventually remove deprecated stuff.

mphe commented:
Ready for review.

limbonaut commented:
> There are `duplicate()`s everywhere in PointArray. I wonder if this is actually necessary. Sometimes those functions are never even used in the project. I can kinda see why a copy would be safer but on the other hand it seems like a needless copy and I would even expect to get the original object in some cases. PointArray even listens for changed signals on each point to make sure they can be safely modified directly. So I think it's safe to remove the duplication and just return the object. What do you think? As for `get_curve()`, I already removed the `duplicate()` in the `PointArray.get_curve()` but kept it in the now deprecated wrapper in shape.gd for compatibility.

Yeah, I think those duplicate() calls can be removed. It's unlikely to cause issues with third-party code, and if a user tempers with those points, they probably expect the changes to apply in the shape state.

limbonaut commented:
Retested, and LGTM :+1: 
And I think you can ditch those dup() calls if you want to.

SirRamEsq commented:
This is looking really good you guys.

I now remember writing all those duplicate() calls. The thought process was, "I'll do this now to make sure everything works, then I'll put some tests in place and slowly phase out the duplicates"
That last part never actually happened :)

I also never really felt the pain of performance myself, because I typically used smaller shapes.

Anyhow, all of this looks like a fantastic improvement.
Thanks so much for all the hard work 🙂
Really exciting stuff.

limbonaut commented:
Checked the additions, and it LGTM :+1:
